strings/kmp/kmp.go

```diff
@@ -1,117 +1,49 @@
-// 		if len(args) <= 2 {
-// 			log.Fatal("Not enough arguments. Two string arguments separated by spaces are required!")
-// 		}
-// 		word = args[1]
-// 		text = args[2]
-// 		for i := 3; i < len(args); i++ {
-// 			text = text + " " + args[i]
-// 		}
-// 	} else { // case of file input
-// 		patFile, err := ioutil.ReadFile("../pattern.txt")
-// 		if err != nil {
-// 			log.Fatal(err)
-// 		}
-// 		textFile, err := ioutil.ReadFile("../text.txt")
-// 		if err != nil {
-// 			log.Fatal(err)
-// 		}
-// 		text = string(textFile)
-// 		word = string(patFile)
-// 	}
+// Kmp Function kmp performing the Knuth-Morris-Pratt algorithm.
+func Kmp(word, text string, patternTable []int) []int {
+	if len(word) > len(text) {
+		return nil
+	}
 
-// 	if len(word) > len(text) {
-// 		log.Fatal("Pattern is longer than text!")
-// 	}
-// 	fmt.Printf("\nRunning: Knuth-Morris-Pratt algorithm.\n\n")
-// 	fmt.Printf("Search word (%d chars long): %q.\n", len(word), word)
-// 	fmt.Printf("Text        (%d chars long): %q.\n\n", len(text), text)
+	var (
+		i, j    int
+		matches []int
+	)
+	for i+j < len(text) {
 
-// 	r := kmp(text, word)
-// 	if r.resultPosition == notFoundPosition {
-// 		fmt.Printf("\n\nWord was not found.\n%d comparisons were done.", r.numberOfComparison)
-// 	} else {
-// 		fmt.Printf("\n\nWord %q was found at position %d in %q. \n%d comparisons were done.", word,
-// 			r.resultPosition, text, r.numberOfComparison)
-// 	}
-// }
+		if word[j] == text[i+j] {
+			j++
+			if j == len(word) {
+				matches = append(matches, i)
 
-// Kmp Function kmp performing the Knuth-Morris-Pratt algorithm.
-// Prints whether the word/pattern was found and on what position in the text or not.
-// m - current match in text, i - current character in w, c - amount of comparisons.
-func Kmp(text string, word string) Result {
-	m, i, c := 0, 0, 0
-	t := kmpTable(word)
-	for m+i < len(text) {
-		fmt.Printf("\n   comparing characters %c %c at positions %d %d", text[m+i], word[i], m+i, i)
-		c++
-		if word[i] == text[m+i] {
-			fmt.Printf(" - match")
-			if i == len(word)-1 {
-				return Result{
-					m, c,
-				}
+				i = i + j
+				j = 0
 			}
-			i++
 		} else {
-			m = m + i - t[i]
-			if t[i] > -1 {
-				i = t[i]
+			i = i + j - patternTable[j]
+			if patternTable[j] > -1 {
+				j = patternTable[j]
 			} else {
-				i = 0
+				j = 0
 			}
 		}
 	}
-	return Result{notFoundPosition,
-		c,
-	}
+	return matches
 }
 
-// Table building algorithm.
-// Takes word to be analyzed and table to be filled.
-func kmpTable(word string) (t []int) {
-	t = make([]int, len(word))
-	pos, cnd := 2, 0
-	t[0], t[1] = -1, 0
-	for pos < len(word) {
-		if word[pos-1] == word[cnd] {
-			cnd++
-			t[pos] = cnd
-			pos++
-		} else if cnd > 0 {
-			cnd = t[cnd]
-		} else {
-			t[pos] = 0
-			pos++
+// Table building for kmp algorithm.
+func Table(w string) []int {
+	var (
+		t []int = []int{-1}
+		k int
+	)
+	for j := 1; j < len(w); j++ {
+		k = j - 1
+		for w[0:k] != w[j-k:j] && k > 0 {
```

Comment:
This feels weird `w[0:k] != w[j-k:j]`

What you're checking is simply, `j == k` isn't it? Why involve pointers when you could just as easily solve the problem without any pointer checks?

Reply:
@tjgurwara99 
Thanks for review.

> w[0:k] ! = w[j-k:j]!

The assumption is that  calculating the number of consecutive matching characters in the string `w`.

The reason for this condition is to compare a string of length k from the beginning of w with a string of length k from the end of `w`.

If they do not match, it subtracts 1 from k and repeats until they match or k=0.

e.g. ) The process flow I envision.

```
w="teste"

// 1. 
j = 5
k = 4
w[0:k] = "test"
w[j-k:j] = "este"

The above does not match, so k = k-1

// 2.
j = 5
k = 3
w[0:k] = "tes"
w[j-k:j] = "ste"

The above does not match, so k = k-1

// 3.
j = 5
k = 2
w[0:k] = "te"
w[j-k:j] = "te"

Match found, so exit from loop as k = 2
```

If I am wrong, or if you find this process complicated, I would appreciate any advice you can give me!

Reply:
Ok, I understand. The following is non-blocking and you can mark this resolved: but I argue that since the condition is part of the core logic not something that should just sit loop conditional (kind of makes it obscure what is happening and difficult to reason about.)

I suggest something like this.

```suggestion
		for k > 0 {
			if w[0:k] != w[j-k:j] {
				break
			}
```

notice that there is no difference but it shows that if condition is the central part of the logic

Reply:
@tjgurwara99 
Thanks for the suggestion.:+1:

This logic seems to accomplish what I was hoping to do.!
But you are right, I think that it difficult to deduce from this logic too.

I thought it might be a good idea to change the logic a little as follows.
I would appreciate your feedback.

```go
// e.g. word = "tested"
// expected: [-1 0 0 1 2 0]
func table(word string) (t []int) {
	t = make([]int, len(word))
	var maxMatchLength int

	t[0] = -1

	for index := 1; index < len(t); index++ {
		// step 2.1: index = 4 is w[4] = "e", maxMatchLength = 1 is w[1] = "e"
		// -> will not get into loop
		//
		// step 3.1: index = 5 is w[5] = "d", maxMatchLength = 2 is w[2] = "s"
		// -> maxMatchLength = 1 is w[1] = "t"
		for maxMatchLength > 0 && word[index] != word[maxMatchLength] {
			maxMatchLength = t[maxMatchLength-1]
		}

		// step 1: index = 3 is w[3] = "t", maxMatchLength = 0 is w[0] = "t"
		// -> maxMatchLength = 1
		//
		// step 2.2: index = 4 is w[4] = "e", maxMatchLength = 1 is w[1] = "e"
		// -> maxMatchLength = 2
		//
		// step 3.2: index = 5 is w[5] = "d", maxMatchLength = 1 is w[1] = "t"
		// -> not match
		if word[index] == word[maxMatchLength] {
			maxMatchLength++
		}
		t[index] = maxMatchLength
	}
	return t
}
```

Reply:
The above still doesn't really help the reader in anyway. I think you should stick to the earlier version (and maybe accept the above recommendation since that is much easier to follow) but this was a non-blocking comment, so I'm approving it nonetheless...